### ISSUE TITLE
Need a policy for removing old packages

### ISSUE THREAD

**Issue by milkypostman** [melpa/melpa#4384]
Given a number of issues about removing old packages lately: #4372, #4382, and #4379 I think we should try to adopt some policy about how we will choose which packages to remove.

I don't feel bad about some error here; a package appears to be old (not updated a lot) and recent downloads are low, then a package gets removed and eventually someone complains and we have to revert the commit. It's not really so bad to have to do this to keep things curated. 

I will work on #4376 and once that is done we should have better stats on what is still actively being used.

**Comment by lunaryorn** [comment]
@milkypostman Mmh, I'm not sure whether you're actually in a position to have a policy.  After all, anyone can just delete a repo from their Github account, and MELPA would end up with nowhere to build from, forcing it to remove the recipe as well, right?

As for the requests that I opened (#4379 and #4380) I _will_ eventually delete these repositories or make them private, whether MELPA decides to remove them or not, for they _are_ unmaintained and outdated, and there's no point keeping them around.  To be honest, I didn't open these PRs to kind of offer a choice of keeping them, but just to raise awareness and give MELPA a grace period before making your builds fail.

**Comment by milkypostman** [comment]
I was thinking about the case prior to repo being deleted. In cases where
the repo still exists but the package is just out of date and not used.

**Comment by lunaryorn** [comment]
@milkypostman Ah, sorry, misunderstood the issue then.  Sorry for the noise.

**Comment by tarsius** [comment]
> I don't feel bad about some error here; a package appears to be old (not updated a lot) and recent downloads are low, then a package gets removed and eventually someone complains and we have to revert the commit. It's not really so bad to have to do this to keep things curated.

So the policy would be "If we feel something isn't maintained anymore and also not used anymore, then we remove it. If it turns out that it actually still used, then we add it back." Seems reasonable and to match what is already being done. I would only add "justification for the removal has to be given in the commit message".

**Comment by xiongtx** [comment]
> So the policy would be "If we feel something isn't maintained anymore and also not used anymore, then we remove it. If it turns out that it actually still used, then we add it back."

Shouldn't the criterion be "does the package still work with the minimum version of Emacs we require"? If a package works fine, but the maintainer hasn't touched it in 2 years, why should it be removed?

**Comment by purcell** [comment]
> Shouldn't the criterion be "does the package still work with the minimum version of Emacs we require"? If a package works fine, but the maintainer hasn't touched it in 2 years, why should it be removed?

Yeah, agree: I'm a bit less gung-ho about deleting old packages. If their upstreams disappear, or if their authors declare them obsolete, then it's worth considering removing them IMO.

**Comment by Boruch-Baum** [comment]
Case in point: I was just skimming the melpa package list and saw package `shell-command`. Its description indicates features that have been standard in emacs for as long as I can remember, but apparently not as long as year 2009, which was the last time the package was updated ... from the emacs orphange repository.

Even if the package functions under current emacs-sen, installing it would add another point of feature-duplication to a user's emacs install, and each point of feature-duplication is a potential point of confusion (eg. "is that an issue in `A` or `B`).

**Comment by tarsius** [comment]
Many of the packages that I moved to the orphanage were moved there because

1. the upstream repository disappeared
2. I want the Emacsmirror to contain all the packages that are in Melpa, and
3. @purcell is "a bit less gung-ho about deleting old packages" (see above).

Some of those packages probably should be removed from Melpa and the Emacsmirror. (In the case of the Emacsmirror, "removing" means moving to the Emacsattic sister-project.)

If someone were to go through the list of packages in the Emacsorphanage, then they would likely find a few packages that should just have been removed instead. If someone volunteers to do that, then they should remove each package using a separate commit in the Melpa repository. The commit message should describe the reason for removal.

Also add a link to `https://github.com/emacsattic/PACKAGE`. That repo doesn't exist at this time yet, but once PACKAGE is removed from Melpa it will also be removed from the orphanage and moved from the mirror to the attic.

**Comment by tarsius** [comment]
As I can [see here](https://emacsmirror.net/stats/emacsorphanage.html#org6923b5c) `shell-command` was added to the orphanage because it was distributed on the Emacs**wiki** and other non-wiki package on Melpa depended on it, so we could not remove it completely when purging the wiki packages.

This [no longer appears to be the case](https://melpa.org/#/shell-command), so this one really should be removed. The package that needed it was `anything`, which has already been removed from Melpa.

**Comment by Boruch-Baum** [comment]
How about also some form of annotation in the packages' summary to indicate their use{ful,less}ness in modern emacsen? It could be in the form of deprecation warning that since emacs version n.m the package's functionality exists in emacs core. You could then extend that feature in the summary page to point to {similar,newer,simpler,more complex} packages.

Slightly off-topic but relevant: there currently exists no mechanics for emacs users to share feedback after evaluating a package. I would absolutely love to able to see either a statistic of how a package has 'scored' over the past year/lifetime. That could that be implemented as commits to a set of html/markdown pages, one for each package...

**Comment by purcell** [comment]
> How about also some form of annotation in the packages' summary to indicate their use{ful,less}ness in modern emacsen? It could be in the form of deprecation warning that since emacs version n.m the package's functionality exists in emacs core. You could then extend that feature in the summary page to point to {similar,newer,simpler,more complex} packages.

I don't think anyone has capacity to maintain such additional metadata.

> there currently exists no mechanics for emacs users to share feedback after evaluating a package.

Yes, I think I noted in the website code that I'd quite like a ratings system. That could still be nice, possibly via a third-party API integration (since we don't run a web app). As for more general comments, I think that would inevitably become a moderation problem.
